### PR TITLE
LJ Hotel

### PR DESCRIPTION
**Pull request by elle-johnnie**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I went back in forth between attempting a BookingSystem module and BookingSystem class. I ended up going the class route since I was exactly sure how to implement mixins .
Describe a concept that you gained more clarity on as you worked on this assignment. | I think I gained more understanding of duck typing. The concept didn't make much sense to me prior to this assignment, but while building it and sending messages between classes I feel like I improved my understanding of the concept.
Describe a _nominal test_ that you wrote for this assignment. | A reservation can be instantiated.
Describe an _edge case test_ that you wrote for this assignment. | An argument error is raised if all rooms are booked.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I did well for the most part. I find it usually helps in the initial building stage of the project. I did find that as I was playing around with block bookings that I felt I just needed to start with code instead of tests since I wasn't sure exactly what I thought it should look like if I approached it first from the test build perspective. I likely would have benefited from pseudocoding this out, but running short on time I jumped straight into building the methods and refactoring others to get block reservations to work.

### CODE REVIEW

File: lib/room.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# contains id, status, cost, and potentially booked dates/availability
+#
+# class methods
+# find all/list all room instances
+require 'csv'
+
+module Hotel
+  class Room
+    attr_reader(:id)
+    attr_accessor(:cost)
```
Comment (CheezItMan) [inline lib/room.rb]: Do you really want `cost` to be an accessor?

File: lib/room.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      20.times do |i|
+        all_rooms << Room.new(id: i + 1, cost: 200.00)
+      end
+
+      return all_rooms
+    end
+
+    # method to find a room
+    def self.find_room(id)
+      @id.find { |room| room.id == id }
```
Comment (CheezItMan) [inline lib/room.rb]: This method doesn't make much sense because your `self.all` method doesn't initialize an `@@id` class variable, and in your `@id` instance variable the instance variable is just a number not a list.

File: lib/reservation.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    # would something like BookedDates = Struct.new(:rsv_start, :rsv_end) work here???
+    # method to make an array of date ranges for reservation
+    # date range up to but not including rsv_end
+    def get_date_range_arr(rsv_start, rsv_end)
+      (rsv_start...rsv_end).to_a.map { |day| day.to_s }
+    end
+
+    def get_total_cost(included_rooms, rsv_start, rsv_end, status)
```
Comment (CheezItMan) [inline lib/reservation.rb:40]: You could do this method with the instance variables instead of using parameters.

Also it's neat how you have a Reservation which can have multiple rooms.

File: lib/booking_system.rb [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      # standardize input date
+      search_date = Date.parse(inspect_date)
+      # select all reservation instances that have booked dates that match
+      # the inspect date
+
+      found_reservations = @reservations.select do |reservation|
+        reservation.booked_dates.include? search_date.to_s
+      end
+      # search through block id's if not found in standard reservations
+      if found_reservations.empty?
```
Comment (CheezItMan) [inline lib/booking_system.rb:74]: Why do this if it doesn't find reservations during that date.  Instead you would **always** want to find both.

File: lib/booking_system.rb [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        found_reservations = @block_reservations.select do |block_res|
+          block_res.booked_dates.include? search_date.to_s
+        end
+      end
+
+      found_reservations = 0 if found_reservations.empty?
```
Comment (CheezItMan) [inline lib/booking_system.rb]: Why return `0` if no reservations?  Why not return an empty list?

File: lib/booking_system.rb [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    # method to find reservation based on id
+    def find_a_reservation(id)
+      reservation = @reservations.find { |rsv| rsv.id == id }
+      raise ArgumentError, 'ID does not exist' if reservation.nil?
+      # implicitly returns 0 if reservation.empty?
+      return reservation
+    end
+
+    # method to return total cost based on reservation id
+    def get_reservation_total(id)
```
Comment (CheezItMan) [inline lib/booking_system.rb:93]: This method seems to duplicate existing functionality.

File: lib/booking_system.rb [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+          unavailable_rooms << rsv.included_rooms
+        end
+      end
+      @block_reservations.each do |rsv|
+        if (rsv.rsv_start > check_in && rsv.rsv_end < check_out) || \
+           (rsv.rsv_start <= check_out && rsv.rsv_end > check_in)
+          unavailable_rooms << rsv.included_rooms
+        end
+      end
+      available_rooms = all_rooms_arr - unavailable_rooms.flatten
```
Comment (CheezItMan) [inline lib/booking_system.rb:118]: Nice!

File: lib/block_reservation.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
+module Hotel
+  class BlockReservation < Reservation
+    attr_reader :group_name, :status, :room_qty
+    def initialize(input)
+      super(input)
+      raise ArgumentError, 'Invalid name' if input[:group_name].nil? || input[:group_name].empty?
+      raise ArgumentError, 'Invalid # of rooms' if input[:room_qty].to_i < 2 || input[:room_qty].to_i > 5
+      @included_rooms = input[:included_rooms]
```
Comment (CheezItMan) [inline lib/block_reservation.rb:9]: You should also be able to make reservations from within a block, after the block is "held."

File: lib/booking_system.rb [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+      return available_rooms
+    end
+
+    def make_a_reservation(guest, check_in, check_out, number_of_rooms = 1, status = :BASE)
+      raise ArgumentError, 'Too many rooms' if number_of_rooms > 20 || !(number_of_rooms.is_a? Integer)
+      open_rooms = find_available_rooms(check_in, check_out).first(number_of_rooms)
```
Comment (CheezItMan) [inline lib/booking_system.rb:125]: What happens if there are not enough available rooms?

File: spec/booking_system_spec.rb [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    it 'can get total cost given a reservation id' do
+      # admin = Hotel::BookingSystem.new
+      id = 2 # 1 room, 1 night, baserate 200.00
+      expect(admin.get_reservation_total(id)).must_equal 1 * 1 * 200.00
+      expect(admin.get_reservation_total(id)).must_be_kind_of Float
+    end
+    it 'can find available rooms given a check in and check out date' do
+      check_in = "October 23 2019"
+      check_out = "October 24 2019"
+      expect(admin.find_available_rooms(check_in, check_out)).must_be_kind_of Array
```
Comment (CheezItMan) [inline spec/booking_system_spec.rb:76]: You could also check and verify each instance is a `Room`

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Good number of commits and good commit messages
Answer comprehension questions	| Check, I'm glad Duck Typing make more sense now.
**Design** |
Each class is responsible for a single piece of the program | Check
Classes are loosely coupled | Check
**Wave 1** |
List rooms | Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check
Invalid date range produces an error | Check
Test coverage | 99.7%
**Wave 2** |
View available rooms for a given date range | Check
Reserving a room that is not available produces an error | **No check for this**
**Wave 3** |
Create a block of rooms | Check
Check if a block has rooms | Check, inherited from `Reservation`
Reserve a room from a block | No way to reserve a room within a block
**Fundamentals** | 
Names variables, classes and modules appropriately | Check
Understanding of variable scope - local vs instance | Check
Can create complex logical structures utilizing variables |Check 
Appropriately uses methods to break down tasks into smaller simpler tasks | Check
Understands the differences between class and instance methods | Check, but see my notes in `room.rb`
Appropriately uses iterators and enumerables | Very well done
Appropriately writes and utilizes classes | Check
Appropriately utilizes modules as a mechanism to encapsulate functionality | Check
**Wrap Up** | **Will complete later**
There is a refactors.txt file | 
The file provides a roadmap to future changes | 
**Additional Feedback** | Very well done.  Nice work.  You hit all the learning goals and got to wave 3.  I like the design and how things are laid out.  See my inline notes, but very well done.  You wrote some really good tests, nice work!
